File: src/main/java/com/uiptv/db/AccountDb.java
package com.uiptv.db;

import com.uiptv.model.Account;
import com.uiptv.util.AccountType;
import com.uiptv.util.PingStalkerPortal;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.List;

import static com.uiptv.db.DatabaseUtils.DbTable.ACCOUNT_TABLE;
import static com.uiptv.db.DatabaseUtils.insertTableSql;
import static com.uiptv.db.DatabaseUtils.updateTableSql;
import static com.uiptv.db.SQLConnection.connect;
import static com.uiptv.util.StringUtils.isBlank;
import static com.uiptv.util.StringUtils.isNotBlank;

public class AccountDb extends BaseDb {
    private static AccountDb instance;


    public static synchronized AccountDb get() {
        if (instance == null) {
            instance = new AccountDb();
        }
        return instance;
    }

    public AccountDb() {
        super(ACCOUNT_TABLE);
    }

    @Override
    Account populate(ResultSet resultSet) {
        Account account = new Account(nullSafeString(resultSet, "accountName"), nullSafeString(resultSet, "username"), nullSafeString(resultSet, "password"), nullSafeString(resultSet, "url"), nullSafeString(resultSet, "macAddress"),nullSafeString(resultSet, "macAddressList"), nullSafeString(resultSet, "serialNumber"), nullSafeString(resultSet, "deviceId1"), nullSafeString(resultSet, "deviceId2"), nullSafeString(resultSet, "signature"), isNotBlank(nullSafeString(resultSet, "type")) ? AccountType.valueOf(nullSafeString(resultSet, "type")) : AccountType.STALKER_PORTAL, nullSafeString(resultSet, "epg"), nullSafeString(resultSet, "m3u8Path"), safeBoolean(resultSet, "pinToTop"));
        account.setDbId(nullSafeString(resultSet, "id"));
        account.setServerPortalUrl(nullSafeString(resultSet, "serverPortalUrl"));
        account.setHttpMethod(isNotBlank(nullSafeString(resultSet, "httpMethod")) ? nullSafeString(resultSet, "httpMethod") : "GET");
        account.setTimezone(isNotBlank(nullSafeString(resultSet, "timezone")) ? nullSafeString(resultSet, "timezone") : "Europe/London");
        return account;
    }

    public List<Account> getAccounts() {
        return getAll("order by pinToTop desc, id", new String[]{});
    }

    public Account getAccountById(String id) {
        return getById(id);
    }

    public Account getAccountByName(String accountName) {
        List<Account> accounts = getAll(" WHERE accountName=?", new String[]{accountName});
        return accounts != null && !accounts.isEmpty() ? accounts.get(0) : null;
    }

    public void save(Account account) {
        Account dbAccount = getAccountByName(account.getAccountName());
        boolean accountExist = dbAccount != null;
        String saveQuery = accountExist ? updateTableSql(ACCOUNT_TABLE) : insertTableSql(ACCOUNT_TABLE);
        try (Connection conn = connect(); PreparedStatement statement = conn.prepareStatement(saveQuery)) {
            statement.setString(1, account.getAccountName());
            statement.setString(2, account.getUsername());
            statement.setString(3, account.getPassword());
            statement.setString(4, account.getUrl());
            statement.setString(5, account.getMacAddress());
            statement.setString(6, account.getMacAddressList());
            statement.setString(7, account.getSerialNumber());
            statement.setString(8, account.getDeviceId1());
            statement.setString(9, account.getDeviceId2());
            statement.setString(10, account.getSignature());
            statement.setString(11, account.getEpg());
            statement.setString(12, account.getM3u8Path());
            statement.setString(13, account.getType().name());
            statement.setString(14, account.getServerPortalUrl());
            statement.setString(15, account.isPinToTop() ? "1" : "0");
            statement.setString(16, account.getHttpMethod());
            statement.setString(17, account.getTimezone());

            if (accountExist) {
                statement.setInt(18, Integer.valueOf(dbAccount.getDbId()));
            }
            statement.execute();
        } catch (SQLException e) {
            throw new RuntimeException("Unable to execute query");
        }
    }

    public void saveServerPortalUrl(Account account) {
        if (isBlank(account.getDbId())) {
            return;
        }
        save(account);
    }

    /**
     * Resolve and persist serverPortalUrl once, so subsequent calls can reuse it.
     * Returns the current/resolved endpoint (possibly blank if resolution failed).
     */
    public String ensureServerPortalUrl(Account account) {
        if (account == null) {
            return "";
        }
        if (isNotBlank(account.getServerPortalUrl())) {
            return account.getServerPortalUrl();
        }
        String resolved = PingStalkerPortal.ping(account);
        if (isNotBlank(resolved)) {
            account.setServerPortalUrl(resolved);
            saveServerPortalUrl(account);
        }
        return account.getServerPortalUrl();
    }
}

File: server/src/main/kotlin/com/uiptv/db/AccountDb.kt
package com.uiptv.db

import com.uiptv.model.Account
import com.uiptv.util.AccountType
import com.uiptv.util.StringUtils.isBlank
import com.uiptv.util.StringUtils.isNotBlank
import java.sql.ResultSet
import java.sql.SQLException

class AccountDb : BaseDb<Account>(DatabaseUtils.DbTable.ACCOUNT_TABLE) {
    companion object {
        private var instance: AccountDb? = null

        @JvmStatic
        @Synchronized
        fun get(): AccountDb {
            if (instance == null) {
                instance = AccountDb()
            }
            return instance!!
        }
    }

    override fun populate(resultSet: ResultSet): Account {
        val type = nullSafeString(resultSet, "type")
        val account = Account(
            nullSafeString(resultSet, "accountName"),
            nullSafeString(resultSet, "username"),
            nullSafeString(resultSet, "password"),
            nullSafeString(resultSet, "url"),
            nullSafeString(resultSet, "macAddress"),
            nullSafeString(resultSet, "macAddressList"),
            nullSafeString(resultSet, "serialNumber"),
            nullSafeString(resultSet, "deviceId1"),
            nullSafeString(resultSet, "deviceId2"),
            nullSafeString(resultSet, "signature"),
            if (isNotBlank(type)) AccountType.valueOf(type!!) else AccountType.STALKER_PORTAL,
            nullSafeString(resultSet, "epg"),
            nullSafeString(resultSet, "m3u8Path"),
            safeBoolean(resultSet, "pinToTop")
        )
        account.dbId = nullSafeString(resultSet, "id")
        account.serverPortalUrl = nullSafeString(resultSet, "serverPortalUrl")
        account.resolveChainAndDeepRedirects = safeBoolean(resultSet, "resolveChainAndDeepRedirects")
        account.httpMethod = if (isNotBlank(nullSafeString(resultSet, "httpMethod"))) nullSafeString(resultSet, "httpMethod")!! else "GET"
        account.timezone = if (isNotBlank(nullSafeString(resultSet, "timezone"))) nullSafeString(resultSet, "timezone")!! else "Europe/London"
        account.xtremeCredentialsJson = nullSafeString(resultSet, "xtremeCredentialsJson")
        return account
    }

    fun getAccounts(): List<Account> = getAll("order by pinToTop desc, id", emptyArray())

    fun getAccountById(id: String): Account? = getById(id)

    fun getAccountByName(accountName: String): Account? = getAll(" WHERE accountName=?", arrayOf(accountName)).firstOrNull()

    fun save(account: Account) {
        val dbAccount = getAccountByName(account.accountName ?: "")
        val accountExist = dbAccount != null
        val saveQuery = if (accountExist) {
            DatabaseUtils.updateTableSql(DatabaseUtils.DbTable.ACCOUNT_TABLE)
        } else {
            DatabaseUtils.insertTableSql(DatabaseUtils.DbTable.ACCOUNT_TABLE)
        }
        try {
            SQLConnection.connect().use { conn ->
                conn.prepareStatement(saveQuery).use { statement ->
                    statement.setString(1, account.accountName)
                    statement.setString(2, account.username)
                    statement.setString(3, account.password)
                    statement.setString(4, account.xtremeCredentialsJson)
                    statement.setString(5, account.url)
                    statement.setString(6, account.macAddress)
                    statement.setString(7, account.macAddressList)
                    statement.setString(8, account.serialNumber)
                    statement.setString(9, account.deviceId1)
                    statement.setString(10, account.deviceId2)
                    statement.setString(11, account.signature)
                    statement.setString(12, account.epg)
                    statement.setString(13, account.m3u8Path)
                    statement.setString(14, account.type.name)
                    statement.setString(15, account.serverPortalUrl)
                    statement.setString(16, if (account.pinToTop) "1" else "0")
                    statement.setString(17, if (account.resolveChainAndDeepRedirects) "1" else "0")
                    statement.setString(18, account.httpMethod)
                    statement.setString(19, account.timezone)
                    if (accountExist) {
                        statement.setInt(20, dbAccount!!.dbId!!.toInt())
                    }
                    statement.execute()
                }
            }
        } catch (e: SQLException) {
            throw DatabaseAccessException("Unable to execute query", e)
        }
    }

    fun saveServerPortalUrl(account: Account) {
        if (isBlank(account.dbId)) {
            return
        }
        save(account)
    }
}
